src/st_carousel.c

```diff
@@ -0,0 +1,223 @@
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#include <math.h>
+
+#include "d_player.h"
+#include "doomdef.h"
+#include "doomtype.h"
+#include "doomstat.h"
+#include "g_nextweapon.h"
+#include "i_timer.h"
+#include "m_array.h"
+#include "m_misc.h"
+#include "r_defs.h"
+#include "st_sbardef.h"
+#include "v_fmt.h"
+#include "v_video.h"
+
+static const char *names[] = {
+    [wp_fist] = "SMFIST",
+    [wp_pistol] = "SMPISG",
+    [wp_shotgun] = "SMSHOT",
+    [wp_chaingun] = "SMMGUN",
+    [wp_missile] = "SMLAUN",
+    [wp_plasma] = "SMPLAS",
+    [wp_bfg] = "SMBFGG",
+    [wp_chainsaw] = "SMCSAW",
+    [wp_supershotgun] = "SMSGN2",
+};
+
+static const weapontype_t weapon_order[] = {
+    wp_fist,
+    wp_chainsaw,
+    wp_pistol,
+    wp_shotgun,
+    wp_supershotgun,
+    wp_chaingun,
+    wp_missile,
+    wp_plasma,
+    wp_bfg,
+};
+
+typedef enum
+{
+    wpi_none = -1,
+    wpi_regular,
+    wpi_selected,
+    wpi_disabled
+} weapon_icon_state_t;
+
+typedef struct
+{
+    weapontype_t weapon;
+    weapon_icon_state_t state;
+} weapon_icon_t;
+
+static weapon_icon_t *weapon_icons;
+static int selected_index = 0;
+
+static int last_index = -1;
+static int last_time;
+static int distance;
+
+static int duration;
+
+void ST_ResetCarousel(void)
+{
+    last_index = -1;
+    last_time = 0;
+    distance = 0;
+    duration = 0;
+}
+
+static void BuildWeaponIcons(const player_t *player)
+{
+    array_clear(weapon_icons);
+
+    for (int i = 0; i < arrlen(weapon_order); ++i)
+    {
+        weapontype_t weapon = weapon_order[i];
+        weapon_icon_state_t state = wpi_none;
+
+        if (last_index == -1 && weapon == player->readyweapon)
+        {
+            last_index = array_size(weapon_icons);
+        }
+
+        if (G_WeaponSelectable(weapon))
+        {
+            if (player->nextweapon == weapon)
+            {
+                selected_index = array_size(weapon_icons);
+                state = wpi_selected;
+            }
+            else if (G_AdjustSelection(weapon) != weapon)
+            {
+                state = wpi_disabled;
+            }
+            else
+            {
+                state = wpi_regular;
+            }
+        }
+        else if (player->weaponowned[weapon])
+        {
+            state = wpi_disabled;
+        }
+
+        if (state != wpi_none)
+        {
+            weapon_icon_t icon = {weapon, state};
+            array_push(weapon_icons, icon);
+        }
+    }
+}
+
+void ST_UpdateCarousel(player_t *player)
+{
+    if (G_NextWeaponActivate())
+    {
+        duration = TICRATE / 2;
+    }
+
+    if (duration == 0)
+    {
+        return;
+    }
+
+    if (automap_on || menuactive || paused || player->playerstate == PST_DEAD)
```

Comment:
Probably should have a "reset if player is spying" check here too. Maybe `player != &players[consoleplayer]` or `displayplayer != consoleplayer`. Something along those lines.

Reply:
Fixed in 5f6b7ee